### PR TITLE
Feature/build system

### PR DESCRIPTION
**Pull request by dierpg**
This pr adds the following:
1. Basic files for the meson build system.
2. Gitignore files.
3. Gtest dependency for unit testing.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#         Luis G. Leon Vega <lleon95@gmail.com>
+#
+#
+
+project(
+  'cynq',
+  'cpp',
+  version : '0.1.0',
+  default_options : ['warning_level=3',
+                     'cpp_std=c++17'])
```
Comment (lleon95) [inline meson.build]: Add the Werror as well

File: meson.options [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#
+# See LICENSE for more information about licensing
+#  Copyright 2023
+#
+# Author: Diego Arturo Avila Torres <dandida95@gmail.com>
+#         Luis G. Leon Vega <lleon95@gmail.com>
+#
+#
+
+option('build_tests_ho', type: 'boolean', value: false)
```
Comment (lleon95) [inline meson.options]: The name of this switch is weird. Use `build-tests` instead

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+## Dependencies definition ##
+#############################
+
+# Directory variables
+common_dependencies_dir = meson.source_root() + '/builddir/common-dependencies'
+
+# Defining Unit Testing dependencies
+gtest_proj = subproject('gtest')
+gtest_dep = gtest_proj.get_variable('gtest_dep')
+gmock_dep = gtest_proj.get_variable('gmock_dep')
```
Comment (lleon95) [inline meson.build]: Remove. These should be only detected when the tests are enabled

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+# options
+
```
Comment (lleon95) [inline meson.build:25]: remove space

File: meson.build [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+# directories
+subdir('src')
+
+# Testing suite
+if build_tests
+  gtest_proj = subproject('gtest')
+  gtest_dep = gtest_proj.get_variable('gtest_dep')
+	gmock_dep = gtest_proj.get_variable('gmock_dep')
```
Comment (lleon95) [inline meson.build]: Use spaces instead of tab

File: meson.build [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  subdir('tests')
```
Comment (lleon95) [inline meson.build:39]: Create the directory with an empty meson.build file. This is going to fail

File: subprojects/.gitignore [hunk 7]
```diff
@@ -0,0 +1,8 @@
+packagecache
+googletest-release-*
+yaml-cpp-yaml-cpp-*
+spdlog-*
+cppzmq-*
+fmt-*
+inih-*
+nlohmann_json-*
```
Comment (lleon95) [inline subprojects/.gitignore]: Are we using all these?

If not, remove this

File: src/.gitignore [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# Prerequisites
```
Comment (lleon95) [inline src/.gitignore]: Remove this file. It is unnecessary

File: meson.build [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  language : 'cpp'
+)
+
+
+#############################
+## Dependencies definition ##
+#############################
+
+# Directory variables
+common_dependencies_dir = meson.source_root() + '/builddir/common-dependencies'
```
Comment (lleon95) [inline meson.build]: Why do you need this?

**Review by lleon95 — CHANGES_REQUESTED** [review]
Please, add placeholder files to compile a shared library. It will help you to define some code that it is not currently defined here.

Moreover, create an examples directory where you can define a small application that links to the shared library so you can check the building system

File: meson.build (hunk 9 above)
Comment (dierpg) [reply meson.build]: it was in the example from isms-kernel, I followed that.

Comment (dierpg) [reply meson.build]: removed

File: meson.build [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+#############################
+
+
+# options
+
+build_tests = get_option('build_tests')
+
+# directories
+projectinc = include_directories('.',
+'src')
```
Comment (lleon95) [inline meson.build]: Headers should be in a directory called "include". Source is only for source code (aka cpp files). So, ideally, this line should be:

```
projectinc = include_directories('include')
```

File: meson.build [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  language : 'cpp'
+)
+
+#############################
+## Dependencies definition ##
+#############################
+
+
+# options
+
```
Comment (lleon95) [inline meson.build:25]: Remove space

File: meson.build [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+#
+# Author: Diego Arturo Avila Torres <dandida95@gmail.com>
+#         Luis G. Leon Vega <luis.leon@ieee.org>
+#
+#
+
+project(
+  'cynq',
+  'cpp',
+  version : '0.1.0',
```
Comment (lleon95) [inline meson.build]: Put this into a single line. Too many lines for this instruction. Save scrolling

File: meson.build [hunk 13]
```diff
@@ -0,0 +1,46 @@
+#
+# See LICENSE for more information about licensing
+#  Copyright 2023
+#
+# Author: Diego Arturo Avila Torres <dandida95@gmail.com>
+#         Luis G. Leon Vega <luis.leon@ieee.org>
+#
+#
```
Comment (lleon95) [inline meson.build]: Remove this last line

File: examples/main.cpp [hunk 14]
```diff
@@ -0,0 +1,13 @@
+/*
+ * See LICENSE for more information about licensing
+ *
+ * Copyright 2023
+ * Author: Luis G. Leon-Vega <luis.leon@ieee.org>
+ *         Diego Arturo Avila Torres <diego.avila@uned.cr>
+ *
+ */
+#include "src/cynq.hpp"
```
Comment (lleon95) [inline examples/main.cpp]: Actually, we should respect the convention. You should have "cynq/cynq.hpp". We don't want to mix the project headers with other headers.

File: meson.build [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+build_tests = get_option('build_tests')
+
+# directories
+projectinc = include_directories('.',
+'src')
+
+
+subdir('examples')
+
+
```
Comment (lleon95) [inline meson.build]: Remove this extra line.

File: src/cynq.hpp [hunk 16]
```diff
@@ -0,0 +1,13 @@
+/*
+ * See LICENSE for more information about licensing
```
Comment (lleon95) [inline src/cynq.hpp]: Move this header to `include/cynq/cynq.hpp`

**Review by lleon95 — CHANGES_REQUESTED** [review]

**Comment by lleon95** [comment]
Please, merge those commits which are targeting an objective. For instance:

* Add placeholder code
* Add building system files

The 22 commits are actually those two.

File: examples/main.cpp [hunk 17]
```diff
@@ -0,0 +1,13 @@
+/*
+ * See LICENSE for more information about licensing
+ *
+ * Copyright 2023
+ * Author: Luis G. Leon-Vega <luis.leon@ieee.org>
+ *         Diego Arturo Avila Torres <diego.avila@uned.cr>
+ *
+ */
+#include "include/cynq/cynq.hpp"
```
Comment (lleon95) [inline examples/main.cpp]: Use `#include <cynq/cynq.hpp>`

**Review by lleon95 — CHANGES_REQUESTED** [review]

**Review by lleon95 — APPROVED** [review]

**Pull request merged**